### ISSUE TITLE
[Skia] TextBox

### ISSUE THREAD

**Issue by jeromelaban** [unoplatform/uno#4319]
Add TextBox support for Skia WPF/GTK

**Comment by microhobby** [comment]
Do we have any work in progress on this, PR or something?

**Comment by microhobby** [comment]
Just to mention that this is a very important component, a basic component. It would be great to have it implemented. Don't get me wrong, I'm not trying to rush the implementation. Just adding my opinion and a vote+ for the this. Thanks for the great work 👍

**Comment by LeftTwixWand** [comment]
Hi. How can I help you to implement TextBox in Skia?

**Comment by Mohsens22** [comment]
Please implement this feature asap, without text input, almost all of Skia heads are useless.

**Comment by francoistanguay** [comment]
> Why is it getting postponed and procrastinated?

Please watch you're tone. This is the first and only warning. 

This has been postponed because we had higher priorities to tackle for our paid customers and/or we invested where we have the most traction and the most users.

We agree that his is important and we're doing our best to contribute as much as possible to our Skia backend.

This is a free and open-source project. It there's something that you want and isn't there. If you want/need it, you have two options, you contribute a PR or you pay for Enterprise Support for us to prioritize.

Thank you

**Comment by jader1313** [comment]
Please, this is essential

**Comment by Mohsens22** [comment]
Hello Uno!
What's the status on this feature?

**Comment by francoistanguay** [comment]
It's shipping in 3.7, within the next 24 hours!

**Comment by LeftTwixWand** [comment]
Uno team, thank you guys so much!

**Comment by jader1313** [comment]
First I would like to thank you for the implementation, but in the initial tests I noticed that the TAB key is not working correctly (next field).

**Comment by MartinZikmund** [comment]
@jader1313 The first version is really initial support, so things like this are not working that perfectly yet. But definitely file issues for all of these and I will tackle them one by one. Keyboard navigation (tabbing) in particular is in active development right now, but having it working with this native text box will be crucial too.

**Comment by mterwoord** [comment]
@MartinZikmund Isn't "really initial support" (almost) done? Textbox works for me in wpf, including being able to copy the contents to clipboard. Multiine works, but `IsReadOnly` doesn't.

**Comment by MartinZikmund** [comment]
@mterwoord yes, many things should be working but things like `IsReadOnly` may not work properly yet. Please file an issue for what is not working yet, and I will try to fill the gaps🙂

**Comment by mmarinchenko** [comment]
@jeromelaban, do you have plans to support .NET Framework targets for WPF head? I tried TextBox on `net472` (supposed to be [fully compatible](https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-implementation-support) with `netstandard2.0`) and found that key input doesn't work (it works on `netcoreapp3.1`, though).

.NET Core runtimes is still not a part of typical Windows installation, so this may be important.

P.S. Alignment for [PlaceholderText](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox.placeholdertext) property also doesn't work.

**Comment by LeftTwixWand** [comment]
@mmarinchenko I think, that is no sense, to do it.
Due to this article: [.NET Framework 4.5.2, 4.6, 4.6.1 will reach End of Support on April 26, 2022](https://devblogs.microsoft.com/dotnet/net-framework-4-5-2-4-6-4-6-1-will-reach-end-of-support-on-april-26-2022/), most of .NET Framework versions will be dead less, than in one year. Release of .NET Framework 4.8 also was the latest release of .NET Framework.

And .NET [roadmap](https://github.com/dotnet/core/blob/main/roadmap.md) shows, that the Microsoft planing to unify their platforms to single .NET platform:
![image](https://user-images.githubusercontent.com/50652041/129798475-1736388d-e1f4-43a3-b486-418b6805764c.png)

I hope, that soon .NET could become a part of default Windows installation. So I think, that it's not relevant to spend time for the old .NET Framework support.
Also, you just can create an app with [Self-Contained runtime](https://docs.microsoft.com/en-us/dotnet/core/deploying/#publish-self-contained) and use [trimming](https://docs.microsoft.com/en-us/dotnet/core/deploying/trim-self-contained) to optimize the size of an installer.

**Comment by mterwoord** [comment]
I hope .NET (core) won't be part of default Windows installation. I can't find the source of it but I read somewhere by MS that's what was the main cause of the standstill of .NET framework. It being part of the OS means very strict rules for compatibility, whereas with .NET (core), they said there will be incompatible changes allowed between LTS and GA releases (ie, going from 6.0 to 7.0 could mean incompatible changes)

**Comment by mmarinchenko** [comment]
@LeftTwixWand, I'm not sure that I got it right... If I understand correctly the main reason why WPF is supported by UNO is Windows 7. So there are my points.

1. Both Win32 and UWP applications for Windows 10+ can be developed using WinUI3/WinAppSDK. So WPF support is not required for current (officially supported) platforms in UNO.

2. The end-of-support logic cannot be applied here because I'm talking about .NET Framework 4.7.2+ specifically. .NET Framework 4.7.2+ will be supported until the day when Microsoft ports all its products to modern .NET (ex. Visual Studio 2022 UI still runs on .NET Framework not .NET 6).

3. In my case, I need to develop a relatively small utility for internal use. I have to target `net472` because we have customers who are still using Windows 7 with extended support from Microsoft.

4. All Windows 7 machines have .NET Framework 4.7.2+ installed since the latest versions of .NET Framework are provided by Microsoft Update service. Modern .NET must be downloaded and installed manually on all machines, which is not an easy task in large organizations.

5. Yes, I may target modern .NET and package my project as self-contained application but then the size of the app will be much bigger and most importantly we'll have to manually rebuild and redistribute the application each time we need to integrate .NET security updates.

I have over options, such as developing this app in pure WPF (or even WinForms), because it's unlikely that it will be needed on other platforms than Windows. But who knows? I have some time before we start so I decided to ask @jeromelaban about UNO plans.

Anyway, thanks for your comment! I hope my intentions are clear now 😅

**Comment by LeftTwixWand** [comment]
@mmarinchenko I'm perfectly understanding you about users on the Windows 7.
We also have about 9% of users on this OS. And it's sad(

**Comment by mmarinchenko** [comment]
@mterwoord I guess the original source is this blog post:

https://devblogs.microsoft.com/dotnet/update-on-net-core-3-0-and-net-framework-4-8/

**Comment by mterwoord** [comment]
If size is of concern, I'd suggest having a look at single file deployment, and linking. Both should be possible with .net 5, and also I think Uno project is also working on getting this working..

**Comment by jeromelaban** [comment]
> @jeromelaban, do you have plans to support .NET Framework targets for WPF head? I tried TextBox on `net472` (supposed to be [fully compatible](https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-implementation-support) with `netstandard2.0`) and found that key input doesn't work (it works on `netcoreapp3.1`, though).
> 
> .NET Core runtimes is still not a part of typical Windows installation, so this may be important.
> 
> P.S. Alignment for [PlaceholderText](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox.placeholdertext) property also doesn't work.

We don't have plans to explicitly support net48 or earlier in the templates because net5 is supported in Win7, but [WPF support nuget packages](https://nuget.info/packages/Uno.UI.Runtime.Skia.Wpf/3.9.7) do have `net47` available. You'll need to adjust the templates created by Uno to to so and use net47 or later to get your app running.

As for the size of the app with net5/net6, you can enable IL trimming, this will make the payload much smaller.

**Comment by mmarinchenko** [comment]
@jeromelaban

> You'll need to adjust the templates created by Uno to to so and use net47 or later to get your app running.

The app runs on `TargetFramework` >= `net472`. The problem is key input in TextBox doesn't work :)

> We don't have plans to explicitly support net48 or earlier...

Ok I see. Thank you all, guys, for your attention!

**Comment by jeromelaban** [comment]
> The app runs on `TargetFramework` >= `net472`. The problem is key input in TextBox doesn't work :)
> 

Can you open an issue on for particular issue with a repro app that shows it? There should not be any different with the behavior from the netcore one.

**Comment by mmarinchenko** [comment]
Yup. #6842

**Comment by MartinZikmund** [comment]
Basic support for text input on Skia is now impelmented, let's create separate issues for bugs and missing features as they are needed.
